File: src/main/kotlin/cz/chobot/container_api/config/security/JWTAuthenticationFilter.java
package cz.chobot.container_api.config.security;


import com.fasterxml.jackson.databind.ObjectMapper;
import cz.chobot.container_api.dto.UserDto;
import cz.chobot.container_api.repository.UserRepository;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Optional;

public class JWTAuthenticationFilter extends UsernamePasswordAuthenticationFilter {
    private AuthenticationManager authenticationManager;

    @Autowired
    private UserRepository userRepository;

    public JWTAuthenticationFilter(AuthenticationManager authenticationManager) {
        this.authenticationManager = authenticationManager;
    }

    public static final String SECRET = "SecretKeyToGenJWTsSecretKeyToGenJWTs";
    public static final long EXPIRATION_TIME = 864_000_000; // 10 days
    public static final String TOKEN_PREFIX = "Bearer ";
    public static final String HEADER_STRING = "Authorization";
    public static final String SIGN_UP_URL = "/api/v1/user";

    @Override
    public Authentication attemptAuthentication(HttpServletRequest req,
                                                HttpServletResponse res) throws AuthenticationException {
        try {
            UserDto creds = new ObjectMapper()
                    .readValue(req.getInputStream(), UserDto.class);

            return authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(
                            creds.getUsername(),
                            creds.getPassword(),
                            new ArrayList<>())
            );
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected void successfulAuthentication(HttpServletRequest req,
                                            HttpServletResponse res,
                                            FilterChain chain,
                                            Authentication auth) throws IOException, ServletException {


        String username = ((User) auth.getPrincipal()).getUsername();
        SecretKey key = Keys.hmacShaKeyFor(SECRET.getBytes());
        String token = Jwts.builder()
                .setSubject(username)
                .signWith(key)
                .compact();
        res.addHeader(HEADER_STRING, TOKEN_PREFIX + token);
        res.addHeader("access-control-expose-headers", "Authorization");
    }
}

File: src/main/kotlin/cz/chobot/container_api/config/security/JWTAuthenticationFilter.kt
package cz.chobot.container_api.config.security


import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.ObjectMapper
import cz.chobot.container_api.dto.UserDto
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.apache.commons.io.IOUtils
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import java.io.IOException
import java.nio.charset.Charset
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class JWTAuthenticationFilter(private val authManager: AuthenticationManager) : UsernamePasswordAuthenticationFilter() {

    init {
        this.setFilterProcessesUrl(SecurityUtil.SIGN_UP_URL)
        this.setAuthenticationManager(authManager)
    }


    /***
     * Check if frontend app is trying to authenticate user. If yes get his username and password from request and check its bcrypt hash against user password bcrypt hash (actual check is done by spring security)
     */
    @Throws(AuthenticationException::class)
    override fun attemptAuthentication(req: HttpServletRequest, res: HttpServletResponse?): Authentication {
        try {
            // map request data to userDto object
            val objectMapper = ObjectMapper()
            objectMapper.configure(JsonParser.Feature.AUTO_CLOSE_SOURCE, true)
            val payload = IOUtils.toString(req.inputStream, Charset.defaultCharset())
            val (username, password) = ObjectMapper().readValue(payload, UserDto::class.java)

            // make spring check login info
            return authManager.authenticate(UsernamePasswordAuthenticationToken(username, password, ArrayList<GrantedAuthority>()))
        } catch (e: IOException) {
            logger.error("Error occurred during login attempt.")
            throw RuntimeException(e)
        }

    }


    /***
     * If loging attempt was successful this method is called. It's generates new JWT token and send it frontend app
     */
    @Throws(IOException::class, ServletException::class)
    override fun successfulAuthentication(req: HttpServletRequest, res: HttpServletResponse, chain: FilterChain?, auth: Authentication) {

        val username = (auth.principal as User).username
        val key = Keys.hmacShaKeyFor(SecurityUtil.SECRET.toByteArray())
        val token = Jwts.builder()
                .setSubject(username)
                .signWith(key)
                .compact()
        res.addHeader(SecurityUtil.HEADER_STRING, SecurityUtil.TOKEN_PREFIX + token)
        res.addHeader("access-control-expose-headers", "Authorization")
    }
}
